src/keys/bip39/mod.rs

```diff
@@ -0,0 +1,750 @@
+        (mnemonic.into_key(), passphrase).into_extended_key()
+    }
+
+    fn into_descriptor_key(
+        self, source: Option<bip32::KeySource>, derivation_path: bip32::DerivationPath
+    ) -> Result<DescriptorKey<Ctx>, KeyError> {
+        let (mnemonic, passphrase) = self;
+
+        (mnemonic.into_key(), passphrase).into_descriptor_key(source, derivation_path)
+    }
+}
+
+#[cfg_attr(docsrs, doc(cfg(feature = "keys-bip39")))]
+impl<Ctx: ScriptContext> DerivableKey<Ctx> for Mnemonic {
+    fn into_extended_key(self) -> Result<ExtendedKey<Ctx>, KeyError> {
+        (self, None).into_extended_key()
+    }
+
+    fn into_descriptor_key(
+        self, source: Option<bip32::KeySource>, derivation_path: bip32::DerivationPath
+    ) -> Result<DescriptorKey<Ctx>, KeyError> {
+        let descriptor_key = self
+            .into_extended_key()?
+            .into_descriptor_key(source, derivation_path)?;
+
+        Ok(set_valid_on_any_network(descriptor_key))
+    }
+}
+
+#[cfg_attr(docsrs, doc(cfg(feature = "keys-bip39")))]
+impl<Ctx: ScriptContext> GeneratableKey<Ctx> for Mnemonic {
+    type Entropy = Box<[u8]>;
+
+    // If word count is `None`, it is picked from the entropy length
+    type Options = (Option<WordCount>, Language);
+    type Error = Option<Error>;
+
+    fn generate_with_entropy(
+        (word_count, language): Self::Options, entropy: Self::Entropy
+    ) -> Result<GeneratedKey<Self, Ctx>, Self::Error> {
+        let entropy = match word_count {
+            Some(count) => &entropy.as_ref()[..(count as usize / 8)],
+            None => &entropy
+        };
+        let mnemonic = Mnemonic::from_entropy_in(language, entropy)?;
+
+        Ok(GeneratedKey::new(mnemonic, any_network()))
+    }
+
+    fn generate(options: Self::Options) -> Result<GeneratedKey<Self, Ctx>, Self::Error> {
+        use rand::{thread_rng, Rng};
+
+        let mut entropy = Box::new([0; 32]);
+        thread_rng().fill(entropy.as_mut());
+        Self::generate_with_entropy(options, entropy)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::str::FromStr;
+    use super::{Language, Mnemonic};
+    use bitcoin::{util::bip32, hashes::hex::FromHex};
+    use crate::keys::{any_network, GeneratableKey, GeneratedKey, bip39::WordCount};
+
+    #[test]
+    fn test_keys_bip39_mnemonic() {
+        let mnemonic =
+            "aim bunker wash balance finish force paper analyst cabin spoon stable organ";
+        let mnemonic = Mnemonic::parse_in(Language::English, mnemonic).unwrap();
+        let path = bip32::DerivationPath::from_str("m/44'/0'/0'/0").unwrap();
+
+        let key = (mnemonic, path);
+        let (desc, keys, networks) = crate::descriptor!(wpkh(key)).unwrap();
+        assert_eq!(desc.to_string(), "wpkh([be83839f/44'/0'/0']xpub6DCQ1YcqvZtSwGWMrwHELPehjWV3f2MGZ69yBADTxFEUAoLwb5Mp5GniQK6tTp3AgbngVz9zEFbBJUPVnkG7LFYt8QMTfbrNqs6FNEwAPKA/0/*)#0r8v4nkv");
+        assert_eq!(keys.len(), 1);
+        assert_eq!(networks.len(), 4);
+    }
+
```

Comment:
I will add test to make sure that I'm catching failing condition for incorrect entropies.


```suggestion
//testing with entropy less than 128 bits
assert_eq!(from_entropy_in(Language::English, vec![b'0'; 15]).unwrap(), Err(Error::InvalidEntropyLength(120)));

//testing with entropy greater than 256 bits
assert_eq!(from_entropy_in(Language::English, vec![b'0'; 33]).unwrap(), Err(Error::InvalidEntropyLength(264)));

//testing entropy which is not a multiple of 32 bits
assert_eq!(from_entropy_in(Language::English, vec![b'0'; 31]).unwrap(), Err(Error::InvalidEntropyLength(248)));

```